src/platform/ESP32/OTAImageProcessorImpl.cpp

```diff
@@ -123,6 +129,96 @@ CHIP_ERROR OTAImageProcessorImpl::ProcessBlock(ByteSpan & block)
     return CHIP_NO_ERROR;
 }
 
+#if CONFIG_ENABLE_DELTA_OTA
+esp_err_t OTAImageProcessorImpl::DeltaOTAReadCallback(uint8_t * buf_p, size_t size, int src_offset)
+{
+    if (size <= 0 || buf_p == NULL)
+    {
+        return ESP_ERR_INVALID_ARG;
+    }
+
+    const esp_partition_t * current_partition = esp_ota_get_running_partition();
+    if (current_partition == NULL)
+    {
+        return ESP_FAIL;
+    }
+
+    esp_err_t err = esp_partition_read(current_partition, src_offset, buf_p, size);
+    ;
+    if (err != ESP_OK)
+    {
+        ESP_LOGE(TAG, "esp_partition_read failed (%s)!", esp_err_to_name(err));
+    }
+
+    return err;
+}
+
+esp_err_t OTAImageProcessorImpl::DeltaOTAWriteHeader(OTAImageProcessorImpl * imageProcessor, const uint8_t * buf_p, size_t size,
+                                                     int index)
+{
+    if (size <= 0 || buf_p == NULL)
+    {
+        return ESP_ERR_INVALID_ARG;
+    }
+
+    static char header_data[IMG_HEADER_LEN];
+    static bool chip_id_verified = false;
+    static int header_data_read  = 0;
+
+    if (!chip_id_verified)
+    {
+        if (header_data_read + size <= IMG_HEADER_LEN)
+        {
+            memcpy(header_data + header_data_read, buf_p, size);
+            header_data_read += size;
+            return ESP_OK;
+        }
+        else
+        {
+            index = IMG_HEADER_LEN - header_data_read;
+            memcpy(header_data + header_data_read, buf_p, index);
+
+            chip_id_verified = true;
+
+            // Write data in header_data buffer.
+            esp_err_t err = esp_ota_write(imageProcessor->mOTAUpdateHandle, header_data, IMG_HEADER_LEN);
+            if (err != ESP_OK)
+            {
+                ESP_LOGE(TAG, "esp_ota_write failed (%s)!", esp_err_to_name(err));
+                return err;
```

Comment:
This is the only place where this function will fail... are we suppose to reset the state of static variables or will it work even if this fails?